### PR TITLE
use Caffeine instead of Guava Cache

### LINKED ISSUE #8399: Migrate from Guava's Cache to Caffeine and prohibit Guava's Cache using forbidden-apis
Guava's Cache is still used in `ResourcePool`, `OnHeapLoadingCache`, and `SegmentLoadDropHandler`.

### PR DESCRIPTION
**Pull request by adrian-wang**
Fixes #8399 .

### Description

Turn to Caffeine instead of Guava's Cache in `ResourcePool`, `OnHeapLoadingCache`, and `SegmentLoadDropHandler`, and interface changes in `OffHeapLoadingCache`.

This PR has:
- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by drcrallen** [comment]
Thanks for the PR! One note: catching blanket `Exception` often breaks `InterruptedException` intentions, especially if the catch just continues on through the loop. Can you please make sure you are more prescriptive on the exceptions caught?

File: extensions-core/lookups-cached-single/src/main/java/org/apache/druid/server/lookup/cache/loading/OnHeapLoadingCache.java [hunk 1]
```diff
@@ -22,29 +22,24 @@
...
 {
   private static final Logger log = new Logger(OnHeapLoadingCache.class);
   private static final int DEFAULT_INITIAL_CAPACITY = 16;
-  //See com.google.common.cache.CacheBuilder#DEFAULT_CONCURRENCY_LEVEL
-  private static final int DEFAULT_CONCURRENCY_LEVEL = 4;
 
   private final Cache<K, V> cache;
   private final AtomicBoolean isClosed = new AtomicBoolean(false);
   @JsonProperty
-  private final int concurrencyLevel;
```
Comment (leventov) [inline extensions-core/lookups-cached-single/src/main/java/org/apache/druid/server/lookup/cache/loading/OnHeapLoadingCache.java]: Please make sure a warning is printed if this legacy JSON property is still specified.

File: extensions-core/lookups-cached-single/src/test/java/org/apache/druid/server/lookup/LoadingLookupTest.java [hunk 2]
```diff
@@ -96,28 +95,6 @@ public void testClose()
     EasyMock.verify(lookupCache, reverseLookupCache);
   }
 
-  @Test
-  public void testApplyWithExecutionError() throws ExecutionException
```
Comment (leventov) [inline extensions-core/lookups-cached-single/src/test/java/org/apache/druid/server/lookup/LoadingLookupTest.java]: Could these tests be retained?

File: extensions-core/lookups-cached-single/src/main/java/org/apache/druid/server/lookup/cache/loading/LoadingCache.java [hunk 3]
```diff
@@ -21,19 +21,18 @@
...
-import java.util.concurrent.ExecutionException;
+import java.util.function.Function;
 
 /**
  * A semi-persistent mapping from keys to values. Cache entries are added using
- * {@link #get(Object, Callable)} and stored in the cache until either evicted or manually invalidated.
+ * {@link #get(Object, Function)} and stored in the cache until either evicted or manually invalidated.
  * <p>
  * <p>Implementations of this interface are expected to be thread-safe, and can be safely accessed
  * by multiple concurrent threads.
```
Comment (leventov) [inline extensions-core/lookups-cached-single/src/main/java/org/apache/druid/server/lookup/cache/loading/LoadingCache.java]: Please ensure a warning is printed when legacy "guava" type is specified, and add a new "onHeap" type.

File: core/src/test/java/org/apache/druid/java/util/common/parsers/TimestampParserTest.java [hunk 4]
```diff
@@ -140,6 +140,7 @@ public void testTimeStampParserWithLongTimeZone()
 
     long millis1 = new DateTime(1994, 11, 9, 4, 0, DateTimeZone.forOffsetHours(-8)).getMillis();
     long millis2 = new DateTime(1994, 11, 9, 4, 0, DateTimeZone.forOffsetHours(-6)).getMillis();
+    // java.util.Locale.setDefault(java.util.Locale.US);
```
Comment (leventov) [inline core/src/test/java/org/apache/druid/java/util/common/parsers/TimestampParserTest.java]: Please remove

File: core/src/test/java/org/apache/druid/java/util/common/parsers/TimestampParserTest.java [hunk 5]
```diff
@@ -157,6 +158,7 @@ public void testTimeStampParserWithLongTimeZone()
   @Test
   public void testTimeZoneAtExtremeLocations()
   {
+    // java.util.Locale.setDefault(java.util.Locale.US);
```
Comment (leventov) [inline core/src/test/java/org/apache/druid/java/util/common/parsers/TimestampParserTest.java]: And this

**Comment by leventov** [comment]
Labelling `Design Review` because the new name for cache property should be vetted  (I suggest "onHeap" not to tie to a specific cache implementation, whether it is Guava or Caffeine).

Labelling `Release Notes` because it entails changes that cluster operators should apply: replace "guava" with "onHeap" (or whatever config name is chosen) and removing the concurrencyLevel property.

**Comment by adrian-wang** [comment]
@leventov @drcrallen Thanks for the review, I'm new to druid, for now I just resolved the conflicts. I'll take time to address your comments, thanks!

**Comment by clintropolis** [comment]
Looks like there is a legit [checkstyle CI failure](https://travis-ci.org/apache/druid/jobs/639561475#L624):

```
[ERROR] /home/travis/build/apache/druid/extensions-core/lookups-cached-single/src/main/java/org/apache/druid/server/lookup/LoadingLookup.java:33:8: Unused import - java.util.concurrent.Callable. [UnusedImports]
```

**Comment by adrian-wang** [comment]
@leventov I've updated the code, now the failed checks are not relevant. can you please re-trigger the CI?

**Comment by ccaominh** [comment]
@adrian-wang I retriggered the jobs for you

**Comment by jihoonson** [comment]
@adrian-wang sorry for the delayed review. Would you please fix the conflicts?

**Comment by adrian-wang** [comment]
@jihoonson  Sorry for the late response, I just merged this with master.

**Comment by adrian-wang** [comment]
gently ping @leventov @jihoonson

**Pull request closed without merging**